cli/internal/run/real_run.go

```diff
@@ -97,10 +104,54 @@ func RealRun(
 		Concurrency: rs.Opts.runOpts.Concurrency,
 	}
 
-	mu := sync.Mutex{}
+	taskCount := len(engine.TaskGraph.Vertices())
+	logChan := make(chan taskLogContext, taskCount)
+	logWaitGroup := sync.WaitGroup{}
+	grouped := rs.Opts.runOpts.LogOrder == "grouped"
+
+	outputLogs := func() {
+		for i := 1; i < taskCount; i++ {
+			logContext := <-logChan
+
+			outBytes := logContext.outBuf.Bytes()
+			errBytes := logContext.errBuf.Bytes()
+
+			_, errOut := os.Stdout.Write(outBytes)
+			_, errErr := os.Stderr.Write(errBytes)
+
+			if errOut != nil || errErr != nil {
+				ec.ui.Error("Failed to output some of the logs.")
+			}
+
+			logWaitGroup.Done()
+		}
+	}
+
+	if grouped {
+		go outputLogs()
+	}
+
+	taskSummaryMutex := sync.Mutex{}
 	taskSummaries := []*runsummary.TaskSummary{}
 	execFunc := func(ctx gocontext.Context, packageTask *nodes.PackageTask, taskSummary *runsummary.TaskSummary) error {
-		taskExecutionSummary, err := ec.exec(ctx, packageTask)
+		logWaitGroup.Add(1)
+		outBuf := &bytes.Buffer{}
+		errBuf := &bytes.Buffer{}
+
+		var outWriter io.Writer
+		var errWriter io.Writer
+
+		if grouped {
+			outWriter = outBuf
+			errWriter = errBuf
+		} else {
+			outWriter = os.Stdout
+			errWriter = os.Stderr
+		}
+
+		ui := concurrentUIFactory.Build(os.Stdin, outWriter, errWriter)
```

Comment:
do we need to pass in `os.Stdin` here? we generally don't handle stdin today so I want to make sure we're not inadvertently adding stdin handling

Reply:
stdin is passed since the underlying ui needs it, this has always been the case. As you can see in
[ui.go:1111](https://github.com/vercel/turbo/pull/3916/files/e34dcb6886a79547b3b0bead02d2e278d79af576#diff-5bf83e5bf3a7f0e8776bac76f967e5e123dbaa43a8a6578d655de332585ad71fL111). The ui factory is simply a wrapper around it that allows me to pass custom writers/readers while maintaining the same nested ui stack (BasicUi > ColoredUi > ...).